IdentityCore/src/webview/MSIDWebviewAuthorization.h

```diff
@@ -26,24 +26,56 @@
 
+typedef void (^MSIDWebviewAuthCompletionHandler)(MSIDWebviewResponse *response, NSError *error);
+
 @interface MSIDWebviewAuthorization : NSObject
 
-+ (id<MSIDWebviewInteracting>)embeddedWebviewControllerWithConfiguration:(MSIDConfiguration *)configuration
-                                                                 factory:(MSIDOAuth2Factory *)factory;
-+ (id<MSIDWebviewInteracting>)embeddedWebviewControllerWithConfiguration:(MSIDConfiguration *)configuration
-                                                                 webview:(WKWebView *)webview
-                                                                 factory:(MSIDOAuth2Factory *)factory;
++ (void)startEmbeddedWebviewAuthWithConfiguration:(MSIDWebviewConfiguration *)configuration
+                                    oauth2Factory:(MSIDOauth2Factory *)oauth2Factory
+                                      verifyState:(BOOL)verifyState
+                                          context:(id<MSIDRequestContext>)context
+                                completionHandler:(MSIDWebviewAuthCompletionHandler)completionHandler;
+
++ (void)startEmbeddedWebviewWebviewAuthWithConfiguration:(MSIDWebviewConfiguration *)configuration
+                                           oauth2Factory:(MSIDOauth2Factory *)oauth2Factory
+                                             verifyState:(BOOL)verifyState
+                                                 webview:(WKWebView *)webview
+                                                 context:(id<MSIDRequestContext>)context
+                                       completionHandler:(MSIDWebviewAuthCompletionHandler)completionHandler;
+
+#if TARGET_OS_IPHONE
++ (void)startSystemWebviewWebviewAuthWithConfiguration:(MSIDWebviewConfiguration *)configuration
+                                         oauth2Factory:(MSIDOauth2Factory *)oauth2Factory
+                                           verifyState:(BOOL)verifyState 
+                                               context:(id<MSIDRequestContext>)context
+                                     completionHandler:(MSIDWebviewAuthCompletionHandler)completionHandler;
+#endif
+
++ (BOOL)setCurrentWebSession:(MSIDWebviewSession *)session;
++ (void)cancelCurrentSession;
 
-+ (id<MSIDWebviewInteracting>)systemWebviewControllerWithConfiguration:(MSIDConfiguration *)configuration
-                                                               factory:(MSIDOAuth2Factory *)factory;
+// This is for system webview auth session on iOS 10 - Thus, a SafariViewController
++ (BOOL)handleURLResponseForSystemWebviewController:(NSURL *)url;
+
+// This can be utilized for having a custom webview controller, and for testing.
++ (void)startSession:(MSIDWebviewSession *)session
+         verifyState:(BOOL)verifyState 
+             context:(id<MSIDRequestContext>)context
+   completionHandler:(MSIDWebviewAuthCompletionHandler)completionHandler;
+
+@property (class, readonly) MSIDWebviewSession *currentSession;
```

Comment:
nit: If you have a class property, can you make it writable and remove setCurrentWebSession?

Reply:
setting session was meant for testing only, I will actually remove it from the header file.

Reply:
Setting it needs to return a BOOL as it can only be set when there is no session running, thus a need for a separate setter.